servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/RetryingHttpRequesterFilter.java

```diff
@@ -226,9 +229,20 @@ public Single<? extends FilterableReservedStreamingHttpConnection> reserveConnec
         @Override
         protected Single<StreamingHttpResponse> request(final StreamingHttpRequester delegate,
                                                         final StreamingHttpRequest request) {
-            final StreamingHttpRequest duplicatedRequest = mayReplayRequestPayload ?
-                    request.transformMessageBody(messageBodyDuplicator()) : request;
-            Single<StreamingHttpResponse> single = delegate.request(duplicatedRequest);
+            // State intentionally outside the defer because the request state is shared across subscribes. If
+            // re-applying operators duplicates logic that isn't desirable and lead to StackOverflowException.
+            final Publisher<Object> originalMessageBody = request.messageBody();
+            final AtomicReference<ContextMap> contextRef = new AtomicReference<>();
+            Single<StreamingHttpResponse> single = Single.defer(() -> {
+                final Single<StreamingHttpResponse> reqSingle = delegate.request(
+                        request.transformMessageBody(mayReplayRequestPayload ?
+                                messageBodyDuplicator(originalMessageBody) : p -> originalMessageBody));
+                final ContextMap map = contextRef.get();
+                return map == null && contextRef.compareAndSet(null, AsyncContext.context()) ?
+                        reqSingle.shareContextOnSubscribe() :
+                        reqSingle.shareContextOnSubscribe(contextRef.get());
```

Comment:
ffti: I think the point is that on follow-up requests we want to reuse the same context entry but why is non-obvious. If there is a short comment about this part it might be helpful, or as likely, a reference to an existing comment somewhere else.